salt/utils/boto3.py

```diff
@@ -0,0 +1,298 @@
+The expected usage is to call `apply_funcs` from the `__virtual__` function
+of the module. This will bring properly initilized partials of  `_get_conn`
+and `_cache_id` into the module's namespace.
+
+Example Usage:
+
+    .. code-block:: python
+
+        import salt.utils.boto3
+
+        def __virtual__():
+            # only required in 2015.2
+            salt.utils.compat.pack_dunder(__name__)
+
+            __utils__['boto.apply_funcs'](__name__, 'vpc')
+
+        def test():
+            conn = _get_conn()
+            vpc_id = _cache_id('test-vpc')
+
+.. versionadded:: 2015.8.0
+'''
+
+# Import Python libs
+from __future__ import absolute_import
+import hashlib
+import logging
+import sys
+from distutils.version import LooseVersion as _LooseVersion  # pylint: disable=import-error,no-name-in-module
+from functools import partial
+
+# Import salt libs
+from salt.ext.six import string_types  # pylint: disable=import-error
+from salt.ext.six.moves import range  # pylint: disable=import-error,redefined-builtin
+from salt.exceptions import SaltInvocationError
+
+# Import third party libs
+# pylint: disable=import-error
+try:
+    # pylint: disable=import-error
+    import boto
+    import boto3
+    import boto.exception
+    import boto3.session
+
+    # pylint: enable=import-error
+    logging.getLogger('boto3').setLevel(logging.CRITICAL)
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+# pylint: enable=import-error
+
+
+log = logging.getLogger(__name__)
+
+
+def __virtual__():
+    '''
+    Only load if boto libraries exist and if boto libraries are greater than
+    a given version.
+    '''
+    # TODO: Determine minimal version we want to support. VPC requires > 2.8.0.
+    required_boto_version = '2.0.0'
+    required_boto3_version = '1.2.1'
+    if not HAS_BOTO:
+        return False
+    elif _LooseVersion(boto.__version__) < _LooseVersion(required_boto_version):
+        return False
+    elif _LooseVersion(boto3.__version__) < _LooseVersion(required_boto3_version):
+        return False
+    else:
+        return True
+
+
+def _option(value):
+    '''
+    Look up the value for an option.
+    '''
+    if value in __opts__:
+        return __opts__[value]
+    master_opts = __pillar__.get('master', {})
+    if value in master_opts:
+        return master_opts[value]
+    if value in __pillar__:
+        return __pillar__[value]
+
+
+def _get_profile(service, region, key, keyid, profile):
+    if profile:
+        if isinstance(profile, string_types):
+            _profile = _option(profile)
+        elif isinstance(profile, dict):
+            _profile = profile
+        key = _profile.get('key', None)
+        keyid = _profile.get('keyid', None)
+        region = _profile.get('region', None)
+
+    if not region and _option(service + '.region'):
+        region = _option(service + '.region')
+
+    if not region:
+        region = 'us-east-1'
```

Comment:
We might want to log here that we have selected this region because one was not passed in.